### PR TITLE
migration: Add case about auto converge

### PR DESCRIPTION
**Pull request by cliping**
VIRT-293939 - VM live migration - guest cpu auto converge(--auto-converge)

Signed-off-by: cliping <lcheng@redhat.com>

### CODE REVIEW

**Comment by cliping** [comment]
Test result:

#avocado run --vt-type libvirt --vt-machine-type q35 migration.migration_efficiency.migration_vm_cpu_auto_converge
No python imaging library installed. Screendump and Windows guest BSOD detection are disabled. In order to enable it, please install python-imaging or the equivalent for your distro.
No python imaging library installed. PPM image conversion to JPEG disabled. In order to enable it, please install python-imaging or the equivalent for your distro.
JOB ID     : 0aa260d3051eebbc6714d6f12b4ccae87491b878
JOB LOG    : /var/lib/avocado/job-results/job-2022-09-05T11.36-0aa260d/job.log
 (1/8) type_specific.io-github-autotest-libvirt.migration.migration_efficiency.migration_vm_cpu_auto_converge.default_throttle.with_precopy.p2p: STARTED
 (1/8) type_specific.io-github-autotest-libvirt.migration.migration_efficiency.migration_vm_cpu_auto_converge.default_throttle.with_precopy.p2p: PASS (297.19 s)
 (2/8) type_specific.io-github-autotest-libvirt.migration.migration_efficiency.migration_vm_cpu_auto_converge.default_throttle.with_precopy.non_p2p: STARTED
 (2/8) type_specific.io-github-autotest-libvirt.migration.migration_efficiency.migration_vm_cpu_auto_converge.default_throttle.with_precopy.non_p2p: PASS (291.73 s)
 (3/8) type_specific.io-github-autotest-libvirt.migration.migration_efficiency.migration_vm_cpu_auto_converge.default_throttle.with_postcopy.p2p: STARTED
 (3/8) type_specific.io-github-autotest-libvirt.migration.migration_efficiency.migration_vm_cpu_auto_converge.default_throttle.with_postcopy.p2p: PASS (291.36 s)
 (4/8) type_specific.io-github-autotest-libvirt.migration.migration_efficiency.migration_vm_cpu_auto_converge.default_throttle.with_postcopy.non_p2p: STARTED
 (4/8) type_specific.io-github-autotest-libvirt.migration.migration_efficiency.migration_vm_cpu_auto_converge.default_throttle.with_postcopy.non_p2p: PASS (293.74 s)
 (5/8) type_specific.io-github-autotest-libvirt.migration.migration_efficiency.migration_vm_cpu_auto_converge.specified_initial_throttle_and_incremental.with_precopy.p2p: STARTED
 (5/8) type_specific.io-github-autotest-libvirt.migration.migration_efficiency.migration_vm_cpu_auto_converge.specified_initial_throttle_and_incremental.with_precopy.p2p: PASS (297.50 s)
 (6/8) type_specific.io-github-autotest-libvirt.migration.migration_efficiency.migration_vm_cpu_auto_converge.specified_initial_throttle_and_incremental.with_precopy.non_p2p: STARTED
 (6/8) type_specific.io-github-autotest-libvirt.migration.migration_efficiency.migration_vm_cpu_auto_converge.specified_initial_throttle_and_incremental.with_precopy.non_p2p: PASS (286.78 s)
 (7/8) type_specific.io-github-autotest-libvirt.migration.migration_efficiency.migration_vm_cpu_auto_converge.specified_initial_throttle_and_incremental.with_postcopy.p2p: STARTED
 (7/8) type_specific.io-github-autotest-libvirt.migration.migration_efficiency.migration_vm_cpu_auto_converge.specified_initial_throttle_and_incremental.with_postcopy.p2p: PASS (297.46 s)
 (8/8) type_specific.io-github-autotest-libvirt.migration.migration_efficiency.migration_vm_cpu_auto_converge.specified_initial_throttle_and_incremental.with_postcopy.non_p2p: STARTED
 (8/8) type_specific.io-github-autotest-libvirt.migration.migration_efficiency.migration_vm_cpu_auto_converge.specified_initial_throttle_and_incremental.with_postcopy.non_p2p: PASS (293.00 s)
RESULTS    : PASS 8 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /var/lib/avocado/job-results/job-2022-09-05T11.36-0aa260d/results.html
JOB TIME   : 2352.88 s

**Comment by cliping** [comment]
@Yingshun  @dzhengfy , please take a look, thanks.

File: provider/migration/migration_base.py [hunk 1]
```diff
@@ -276,3 +277,53 @@ def check_qemu_mem_lock_hard_limit(params):
     output = libvirt_memory.get_qemu_process_memlock_hard_limit()
     if int(output) != int(expect_hard_limit) * 1024:
         raise exceptions.TestFail("'%s' is not matched expect hard limit '%s'" % (output, expect_hard_limit))
+
+
+def check_auto_converge_during_mig(params):
+    initial = int(params.get("initial_throttle"))
```
Comment (Yingshun) [inline provider/migration/migration_base.py:283]: Please update docstring for this function and it's parameters and exceptions. Thanks!

**Review by Yingshun — COMMENTED** [review]
Others LGTM

File: provider/migration/migration_base.py (hunk 1 above)
Comment (cliping) [reply provider/migration/migration_base.py:283]: Added. Thanks.

**Review by dzhengfy — COMMENTED** [review]
Others LGTM

File: libvirt/tests/src/migration/migration_efficiency/migration_efficiency.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    :param test: test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+    def setup_auto_converge():
+        """
+        Setup for auto converge
+
+        """
+        test.log.info("Setup maxdowntime.")
```
[the remark this answers is not in the document]
Comment (cliping) [reply libvirt/tests/src/migration/migration_efficiency/migration_efficiency.py:19]: Updated, thanks.

**Review by dzhengfy — APPROVED** [review]
LGTM

**Review by Yingshun — APPROVED** [review]

**Pull request merged**